src/ntpclient.cpp

```diff
@@ -0,0 +1,195 @@
+
+        ip::udp::endpoint receiver_endpoint = *resolver.resolve(query);
+
+        ip::udp::socket socket(io_service);
+        socket.open(ip::udp::v4());
+
+        boost::array<unsigned char, 48> sendBuf = {{010,0,0,0,0,0,0,0,0}};
+
+        socket.send_to(boost::asio::buffer(sendBuf), receiver_endpoint);
+
+        boost::array<unsigned long, 1024> recvBuf;
+        ip::udp::endpoint sender_endpoint;
+
+        try
+        {
+
+            fd_set fileDescriptorSet;
+            struct timeval timeStruct;
+
+            // set the timeout to 10 seconds
+            timeStruct.tv_sec = GetArg("-ntptimeout", 10);
+            timeStruct.tv_usec = 0;
+            FD_ZERO(&fileDescriptorSet);
+
+            int nativeSocket = socket.native();
+
+            FD_SET(nativeSocket,&fileDescriptorSet);
+
+            select(nativeSocket+1,&fileDescriptorSet,NULL,NULL,&timeStruct);
+
+            if(!FD_ISSET(nativeSocket,&fileDescriptorSet))
+            {
+
+                LogPrint("ntp", "[NTP] Could not read socket from NTP server %s (Read timeout)\n", sHostName);
+
+            }
+            else
+            {
+
+                socket.receive_from(boost::asio::buffer(recvBuf), sender_endpoint);
+                timeRecv = ntohl((time_t)recvBuf[4]);
+                timeRecv-= 2208988800U;  // Substract 01/01/1970 == 2208988800U
+
+                LogPrint("ntp", "[NTP] Received timestamp: %ll \n", (uint64_t)timeRecv);
+
+            }
+
+        }
+        catch (std::exception& e)
+        {
+
+             LogPrintf("[NTP] Could not read clock from NTP server %s (%s)\n", sHostName, e.what());
+
+        }
+
+    }
+    catch (std::exception& e)
+    {
+
+        LogPrintf("[NTP] Could not open socket to NTP server %s (%s)\n", sHostName, e.what());
+
+    }
+
+    return (int64_t)timeRecv;
+}
+
+bool NtpClockSync() {
+    LogPrintf("[NTP] Starting clock sync...\n");
+
+    std::vector<std::string> vNtpServers;
+    std::vector<int64_t> vResults;
+
+    if (mapMultiArgs["-ntpservers"].size() > 0)
+    {
+        vNtpServers = mapMultiArgs["-ntpservers"];
+    }
+    else
+    {
+        vNtpServers = vDefaultNtpServers;
+    }
+
+    string sReport = "";
+    string sPrevServer = "";
+    int64_t nPrevMeasure = -1;
+
+    random_shuffle(vNtpServers.begin(), vNtpServers.end(), GetRandInt);
+
+    for(unsigned int i = 0; i < vNtpServers.size(); i++)
+    {
+        string s = vNtpServers[i];
+        CNtpClient ntpClient(s);
+        int64_t nTimestamp = ntpClient.getTimestamp();
+        if(nTimestamp > -1)
+        {
+            int64_t nClockDrift = GetTimeNow() - nTimestamp;
+
+            // We push if its the first entry
+            if(vResults.size() == 0)
+            {
+                vResults.push_back(nClockDrift);
+
+                string sign = "";
+                if(nClockDrift > 0)
+                    sign = "+";
+                else if (nClockDrift < 0)
```

Comment:
can nClockDrift == 0 ? if so, the sign variable won't be defined with this logic.

Reply:
sign is "" if nClockDrift is 0.

sign is just the prefix (string) to prepend to the drift when shown on the log file